### PR TITLE
support get cpu memory info from env

### PR DESCRIPTION
**Pull request by weeds085490**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

To run  clickhouse in docker, we get the memory and CPU information of the physical machine instead of 
Information after cgroups restrictions when calling `getMemoryAmount` or `getNumberOfPhysicalCPUCores`(sysctl is called). Here provide a way to get this information from environment variables.

Detailed description / Documentation draft:

To get cpu and memory information from env, we should add the `cmake` flags  `-DCLICKHOUSE_HARDWARE_INFO_FROM_ENV=ON`

Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by amosbird** [comment]
> To run clickhouse in docker

Perhaps we can extend those functions so that they are aware of containers?

**Comment by weeds085490** [comment]
> > To run clickhouse in docker
> 
> Perhaps we can extend those functions so that they are aware of containers?

In my opinion, Obtaining resource information from environment variables is only a way to avoid getting information about the physical/virtual machine.  There are others:

 * From /proc (Using `lxcfs`)
 * Using namespaced `sysctl`(Some linux kernal already supports this feature).
 
This is just to give users a way to get information they they expected.

**Comment by amosbird** [comment]
> > > To run clickhouse in docker
> > 
> > Perhaps we can extend those functions so that they are aware of containers?
> 
> @amosbird  
> In my opinion, Obtaining resource information from environment variables is only a way to avoid getting information about the physical/virtual machine.  There are others:
> 
>  * From /proc (Using `lxcfs`)
>  * Using namespaced `sysctl`(Some linux kernal already supports this feature).
>  
> This is just to give users a way to get information they they expected.

What's the problem of using /proc ? And I feel that it's better to add server options instead of relying on environ

**Comment by weeds085490** [comment]
> > > > To run clickhouse in docker
> > > 
> > > 
> > > Perhaps we can extend those functions so that they are aware of containers?
> > 
> > 
> > @amosbird
> > In my opinion, Obtaining resource information from environment variables is only a way to avoid getting information about the physical/virtual machine.  There are others:
> > 
> > * From /proc (Using `lxcfs`)
> > * Using namespaced `sysctl`(Some linux kernal already supports this feature).
> > 
> > This is just to give users a way to get information they they expected.
> 
> What's the problem of using /proc ? And I feel that it's better to add server options instead of relying on environ


> > > > To run clickhouse in docker
> > > 
> > > 
> > > Perhaps we can extend those functions so that they are aware of containers?
> > 
> > 
> > @amosbird
> > In my opinion, Obtaining resource information from environment variables is only a way to avoid getting information about the physical/virtual machine.  There are others:
> > 
> > * From /proc (Using `lxcfs`)
> > * Using namespaced `sysctl`(Some linux kernal already supports this feature).
> > 
> > This is just to give users a way to get information they they expected.
> 
> What's the problem of using /proc ? And I feel that it's better to add server options instead of relying on environ

`/proc` + `lxcfs` is indeed a better way. I will also try to provide options to get all the information from `/proc` later. But obtaining from environment variables is very straightforward and simple. In some cases, we can use it directly.

**Comment by weeds085490** [comment]
> > > > > To run clickhouse in docker
> > > > 
> > > > 
> > > > Perhaps we can extend those functions so that they are aware of containers?
> > > 
> > > 
> > > @amosbird
> > > In my opinion, Obtaining resource information from environment variables is only a way to avoid getting information about the physical/virtual machine.  There are others:
> > > 
> > > * From /proc (Using `lxcfs`)
> > > * Using namespaced `sysctl`(Some linux kernal already supports this feature).
> > > 
> > > This is just to give users a way to get information they they expected.
> > 
> > 
> > What's the problem of using /proc ? And I feel that it's better to add server options instead of relying on environ
> 
> > > > > To run clickhouse in docker
> > > > 
> > > > 
> > > > Perhaps we can extend those functions so that they are aware of containers?
> > > 
> > > 
> > > @amosbird
> > > In my opinion, Obtaining resource information from environment variables is only a way to avoid getting information about the physical/virtual machine.  There are others:
> > > 
> > > * From /proc (Using `lxcfs`)
> > > * Using namespaced `sysctl`(Some linux kernal already supports this feature).
> > > 
> > > This is just to give users a way to get information they they expected.
> > 
> > 
> > What's the problem of using /proc ? And I feel that it's better to add server options instead of relying on environ
> 
> `/proc` + `lxcfs` is indeed a better way. I will also try to provide options to get all the information from `/proc` later. But obtaining from environment variables is very straightforward and simple. In some cases, we can use it directly.

Just like we can get config from `env`, `zookeeper`, `file`...  Actually  it is just a simple way

File: src/Common/getNumberOfPhysicalCPUCores.cpp [hunk 1]
```diff
@@ -9,6 +9,15 @@
 
 unsigned getNumberOfPhysicalCPUCores()
 {
+#if CLICKHOUSE_HARDWARE_INFO_FROM_ENV
+    const char * raw_env_cores = std::getenv("CLICKHOUSE_CPU_CORES");
```
Comment (azat) [inline src/Common/getNumberOfPhysicalCPUCores.cpp:13]: Maybe it is better to provide a config directive and use `from_env` modifier in config if you need to override it from env?

i.e.

```xml
<yandex>
    <cpu_cores from_env="CLICKHOUSE_CPU_CORES" />
</yandex>
```

And I think that detecting it automatically should be done in a first place (as already suggested by @amosbird ), this option should be used only if automatic detection does not work correctly.

Comment (weeds085490) [reply src/Common/getNumberOfPhysicalCPUCores.cpp:13]: Ok.I will refine it later.

**Comment by filimonov** [comment]
Any updates here?

See also:
#16918
#2261 + #2325 
#18448
#17556
#20608

**Comment by alexey-milovidov** [comment]
@mergifyio update

File: base/common/CMakeLists.txt [hunk 2]
```diff
@@ -35,6 +35,12 @@ endif ()
 
 add_library (common ${SRCS})
 
+if (CLICKHOUSE_HARDWARE_INFO_FROM_ENV)
```
Comment (alexey-milovidov) [inline base/common/CMakeLists.txt:38]: I don't like build options, let's remove them.

**Comment by alexey-milovidov** [comment]
> And I think that detecting it automatically should be done in a first place

+

**Pull request closed without merging**